### PR TITLE
Added options to name default routes differently

### LINKED ISSUE #70: Add option to pluralize the default api routes [20 + 10]
- [ ] Refactor the default api routes to be lower cased and dash-separated (see [our api guidelines](https://github.com/hpi-swa-teaching/RatPack2.0/wiki/Best-Practices-for-API-Design))
- [ ] Setting for whether api routes should be created with the pluralized version of the model name
- [ ] Allow definition of the api route in the model. If this is set, it should override the automatic route names

### PR DESCRIPTION
**Pull request by schnaaabeltier**
Fix #70

### CODE REVIEW

File: packages/RatPack.package/RPApplication.class/class/routeNamingStrategies.st [hunk 1]
```diff
@@ -0,0 +1,4 @@
+constants
+routeNamingStrategies 
+
+	^ OrderedCollection  newFrom: { RPPluralizer new. RPSeparator newWithSeparator: '_'. RPLowercaser new. }
```
Comment (richardebeling) [inline packages/RatPack.package/RPApplication.class/class/routeNamingStrategies.st]: I would disable pluralization by default.

File: packages/RatPack.package/RPApplication.class/instance/getRegisteredObjectsFromRequest..st [hunk 2]
```diff
@@ -1,8 +1,9 @@
 endpoint
-getRegisteredObjectsFromRequest: aRPModelname
+getRegisteredObjectsFromRequest: aRoute
  
       | modelObjects string aRPModelClass |
-	aRPModelClass := self getRegisteredModel: aRPModelname.
+	
+	aRPModelClass := self getModelForRoute: aRoute.
       modelObjects := self database loadModelObjectsOf: aRPModelClass.  
 	string := '{"data":[', ((modelObjects collect: [:each |
```
Comment (richardebeling) [inline packages/RatPack.package/RPApplication.class/instance/getRegisteredObjectsFromRequest..st]: indentation looks messy here

File: packages/RatPack.package/RPApplication.class/instance/getRegisteredObjectsFromRequestAsHtml..st [hunk 3]
```diff
@@ -1,8 +1,8 @@
 endpoint
-getRegisteredObjectsFromRequestAsHtml: aRPModelname
+getRegisteredObjectsFromRequestAsHtml: aRoute
  
       | objectString objects |
 	objectString := ''.
-      objects := self database loadModelObjectsOf: (self getRegisteredModel: aRPModelname).
+      objects := self database loadModelObjectsOf: (self getModelForRoute: aRoute).
```
Comment (richardebeling) [inline packages/RatPack.package/RPApplication.class/instance/getRegisteredObjectsFromRequestAsHtml..st]: indentation

File: packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st [hunk 4]
```diff
@@ -3,4 +3,8 @@ initializeRoutes
 
 	routes := IdentityDictionary new.
 	self initializePragmaRoutes.
-	self initializeEndpointRoutes 	
\ No newline at end of file
+	self initializeEndpointRoutes. 	
+		
+	self defaultRoutesModelLookupTable: Dictionary new.
```
Comment (richardebeling) [inline packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st:8]: Doesn't initializeEndpointRoutes access this dictionary and write entries to it? Why is this correct?

File: packages/RatPack.package/RPSeparator.class/instance/nameRoute..st [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+	separatorPositions := OrderedCollection new.
+	1 to: aClassName size do: [ :index | 
+		char := aClassName at: index.
+		(index + 2 <= aClassName size ) ifTrue: [
+			nextChar := (aClassName at: (index + 1)) asCharacter.
+			secondNextChar := (aClassName at: (index + 2)) asCharacter.
+			((nextChar isUppercase and: [ secondNextChar isLowercase. ]) or: 
+			(nextChar isUppercase and: [ secondNextChar isUppercase and: [char isLowercase ]] )) ifTrue: [
+				separatorPositions add: index. ]. ]. ].
```
Comment (richardebeling) [inline packages/RatPack.package/RPSeparator.class/instance/nameRoute..st]: Could we put finding the seperator positions in an own method? 25 lines is pretty long

File: packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st [hunk 6]
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
+routeNamingStrategies 
+
+	^ OrderedCollection newFrom: { RPTestRouteNamingStrategy new. }.
```
Comment (richardebeling) [inline packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st]: @tom95 suggested `^ { RPTestRouteNamingStrategy new. }.`

File: packages/RatPack.package/RPTests.class/instance/testDefaultRouteNamingStrategies.st [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+	self assert: ((app routes at: #get:) anySatisfy: [ :route | route key = (routeName, RPApplication wildcard)]).
+	self assert: ((app routes at: #put:) anySatisfy: [ :route | route key = (routeName, RPApplication wildcard)]).
+	self assert: ((app routes at: #post:) anySatisfy: [ :route | route key = routeName]).
+	self assert: ((app routes at: #delete:) anySatisfy: [ :route | route key = (routeName, RPApplication wildcard)]).
+	
+	routeName := RPApplication htmlRoutePrefix, (RPTestRouteNamingStrategy new nameRoute: RPTestModel name).
+	
+	self assert: ((app routes at: #get:) anySatisfy: [ :route | route key = routeName]).
+	self assert: ((app routes at: #get:) anySatisfy: [ :route | route key = (routeName, RPApplication wildcard)]).
+
```
Comment (richardebeling) [inline packages/RatPack.package/RPTests.class/instance/testDefaultRouteNamingStrategies.st]: This and testDefaultRouteNameInModel.st share a lot of code. Could we refactor the duplications in an own method?

File: packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st (hunk 6 above)
Comment (schnaaabeltier) [reply packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st]: It has to be ordered (if you have more than one strategy different execution order leads to different results)

File: packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st (hunk 4 above)
Comment (schnaaabeltier) [reply packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st:8]: This is used to map route name to model name so we can access models which have different names than their routes. `initializeEndpointRoutes` does not access it

File: packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st (hunk 6 above)
Comment (richardebeling) [reply packages/RatPack.package/RPTestApp.class/class/routeNamingStrategies.st]: `^ { RPTestRouteNamingStrategy new. } asOrderedCollection` then, sorry. I don't really care about this, it's just what was suggested to us in the meeting last week.

Edit: Actually, isn't the order also guaranteed to be kept as long as you don't modify the collection? For documentation sake we can use a ordered collection, though.

File: packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st (hunk 4 above)
Comment (richardebeling) [reply packages/RatPack.package/RPApplication.class/instance/initializeRoutes.st:8]: Ah I see

File: packages/RatPack.package/RPApplication.class/instance/postFromRequest..st [hunk 8]
```diff
@@ -5,4 +5,4 @@ postFromRequest: aRoute
 	aRPModelClass := self getModelForRoute: aRoute.
 	jsonString := self request rawPostContents.
 	responseId := self persistObjectBy: jsonString from: aRPModelClass.
-	^ self getObjectFromRequest: aRPModelname with: responseId.
+	^ self getObjectFromRequest: aRoute with: responseId.
```
Comment (richardebeling) [inline packages/RatPack.package/RPApplication.class/instance/postFromRequest..st]: Could you change this so it's called `apiRouteSuffix` or something similar to explicitly show that we mean everything after the `/api/` part? This applies to all methods that handle the route in some way where it's actually just a suffix.

**Pull request merged**